File: python_scripts/chart_generator.py
#!/usr/bin/env python3
"""
Chart Generator for 10 Different Chart Types
Using matplotlib and seaborn
"""

import sys
import json
import pandas as pd
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
from matplotlib import rcParams

# Set style
sns.set_style("whitegrid")
rcParams['figure.facecolor'] = 'white'
rcParams['axes.facecolor'] = 'white'

def create_chart(chart_type, df, params):
    """Create chart based on type and parameters"""
    
    try:
        # Create figure
        fig, ax = plt.subplots(figsize=(10, 6))
        
        x_col = params.get('xColumn')
        y_col = params.get('yColumn')
        group_col = params.get('groupColumn')
        title = params.get('title', f'{chart_type.capitalize()} Chart')
        
        if chart_type == 'boxplot':
            # 1. BOXPLOT - Distribution and outliers
            if group_col:
                sns.boxplot(data=df, x=x_col, y=y_col, palette='viridis', ax=ax)
            else:
                sns.boxplot(data=df, y=y_col, color='#667eea', ax=ax)
            ax.set_xlabel(x_col if x_col else '')
            ax.set_ylabel(y_col)
            plt.xticks(rotation=45, ha='right')
            
        elif chart_type == 'scatter':
            # 2. SCATTER PLOT - Correlation between two variables
            if group_col:
                for group in df[group_col].unique():
                    group_data = df[df[group_col] == group]
                    ax.scatter(group_data[x_col], group_data[y_col], 
                             label=group, alpha=0.7, s=50)
                ax.legend()
            else:
                ax.scatter(df[x_col], df[y_col], color='#667eea', alpha=0.7, s=50)
            ax.set_xlabel(x_col)
            ax.set_ylabel(y_col)
            
        elif chart_type == 'line':
            # 3. LINE CHART - Time series and trends
            if group_col:
                for group in df[group_col].unique():
                    group_data = df[df[group_col] == group]
                    ax.plot(group_data[x_col], group_data[y_col], 
                           label=group, marker='o', linewidth=2)
                ax.legend()
            else:
                ax.plot(df[x_col], df[y_col], color='#667eea', 
                       marker='o', linewidth=2, markersize=6)
            ax.set_xlabel(x_col)
            ax.set_ylabel(y_col)
            
        elif chart_type == 'bar':
            # 4. BAR CHART - Categorical comparisons
            if group_col:
                df_pivot = df.pivot_table(values=y_col, index=x_col, 
                                         columns=group_col, aggfunc='sum')
                df_pivot.plot(kind='bar', ax=ax, colormap='viridis')
                ax.legend(title=group_col)
            else:
                df_agg = df.groupby(x_col)[y_col].sum()
                ax.bar(df_agg.index, df_agg.values, color='#667eea', alpha=0.8)
            ax.set_xlabel(x_col)
            ax.set_ylabel(y_col)
            plt.xticks(rotation=45, ha='right')
            
        elif chart_type == 'histogram':
            # 5. HISTOGRAM - Frequency distribution
            ax.hist(df[x_col].dropna(), bins=30, color='#667eea', 
                   edgecolor='#f093fb', alpha=0.7)
            ax.set_xlabel(x_col)
            ax.set_ylabel('Frequency')
            
        elif chart_type == 'violin':
            # 6. VIOLIN PLOT - Distribution density
            if group_col:
                sns.violinplot(data=df, x=x_col, y=y_col, palette='viridis', ax=ax)
            else:
                sns.violinplot(data=df, y=y_col, color='#667eea', ax=ax)
            ax.set_xlabel(x_col if x_col else '')
            ax.set_ylabel(y_col)
            plt.xticks(rotation=45, ha='right')
            
        elif chart_type == 'density':
            # 7. DENSITY PLOT - Continuous probability distribution
            if group_col:
                for group in df[group_col].unique():
                    group_data = df[df[group_col] == group]
                    group_data[x_col].plot(kind='density', label=group, ax=ax)
                ax.legend()
            else:
                df[x_col].plot(kind='density', color='#667eea', ax=ax)
            ax.set_xlabel(x_col)
            ax.set_ylabel('Density')
            
        elif chart_type == 'heatmap':
            # 8. HEATMAP - Correlation matrix
            numeric_cols = df.select_dtypes(include=[np.number])
            corr_matrix = numeric_cols.corr()
            sns.heatmap(corr_matrix, annot=True, fmt='.2f', 
                       cmap='coolwarm', center=0, ax=ax)
            plt.xticks(rotation=45, ha='right')
            plt.yticks(rotation=0)
            
        elif chart_type == 'ridgeline':
            # 9. RIDGELINE PLOT - Multiple distribution comparison
            # Simplified version using violin plots
            if group_col:
                sns.violinplot(data=df, x=y_col, y=x_col, palette='viridis', ax=ax)
            else:
                # Create categories if not provided
                df['category'] = pd.cut(df[x_col], bins=5)
                sns.violinplot(data=df, x=y_col, y='category', palette='viridis', ax=ax)
            ax.set_xlabel(y_col)
            ax.set_ylabel(x_col)
            
        elif chart_type == 'pie':
            # 10. PIE CHART - Proportions and percentages
            pie_data = df[x_col].value_counts()
            colors = plt.cm.viridis(np.linspace(0, 1, len(pie_data)))
            ax.pie(pie_data.values, labels=pie_data.index, autopct='%1.1f%%',
                  colors=colors, startangle=90)
            ax.axis('equal')
            
        else:
            raise ValueError(f"Unknown chart type: {chart_type}")
        
        # Set title
        ax.set_title(title, fontsize=16, fontweight='bold', pad=20)
        
        # Adjust layout
        plt.tight_layout()
        
        # Save figure
        output_path = params['outputPath']
        plt.savefig(output_path, dpi=300, bbox_inches='tight', 
                   facecolor='white', edgecolor='none')
        plt.close()
        
        print(f"Chart created successfully: {output_path}")
        
    except Exception as e:
        print(f"Error creating chart: {str(e)}", file=sys.stderr)
        raise

def main():
    """Main function"""
    if len(sys.argv) < 2:
        print("No parameters file provided!", file=sys.stderr)
        sys.exit(1)
    
    try:
        # Read parameters from file
        params_file = sys.argv[1]
        with open(params_file, 'r', encoding='utf-8') as f:
            params = json.load(f)
        
        # Read data
        with open(params['dataPath'], 'r', encoding='utf-8') as f:
            data = json.load(f)
        
        # Convert to DataFrame
        df = pd.DataFrame(data)
        
        # Create chart
        create_chart(params['chartType'], df, params)
        
    except Exception as e:
        print(f"Error: {str(e)}", file=sys.stderr)
        sys.exit(1)

if __name__ == '__main__':
    main()

File: r_scripts/chart_generator.R
#!/usr/bin/env Rscript

# Chart Generator for 10 Different Chart Types
# Using ggplot2 and related packages

library(ggplot2)
library(dplyr)
library(tidyr)
library(jsonlite)
library(ggridges)
library(viridis)
library(reshape2)
library(RColorBrewer)

# Parse command line arguments
args <- commandArgs(trailingOnly = TRUE)
if (length(args) == 0) {
  stop("No parameters provided!")
}

# Parse JSON parameters
params <- fromJSON(args[1])

# Read data
data <- fromJSON(params$dataPath)

# Convert to data frame
df <- as.data.frame(data)

# Set default theme
theme_set(theme_minimal() + 
  theme(
    plot.title = element_text(size = 16, face = "bold", hjust = 0.5),
    plot.background = element_rect(fill = "white", color = NA),
    panel.background = element_rect(fill = "white", color = NA)
  ))

# Chart generation functions
create_chart <- function(chart_type, df, params) {
  
  tryCatch({
    
    if (chart_type == "boxplot") {
      # 1. BOXPLOT - Distribution and outliers
      p <- ggplot(df, aes_string(x = params$xColumn, y = params$yColumn)) +
        geom_boxplot(fill = "#667eea", alpha = 0.7, outlier.color = "#f093fb") +
        labs(title = params$title, x = params$xColumn, y = params$yColumn) +
        theme(axis.text.x = element_text(angle = 45, hjust = 1))
      
    } else if (chart_type == "scatter") {
      # 2. SCATTER PLOT - Correlation between two variables
      if (!is.null(params$groupColumn) && params$groupColumn != "") {
        p <- ggplot(df, aes_string(x = params$xColumn, y = params$yColumn, color = params$groupColumn)) +
          geom_point(size = 3, alpha = 0.7) +
          scale_color_viridis(discrete = TRUE) +
          labs(title = params$title, x = params$xColumn, y = params$yColumn)
      } else {
        p <- ggplot(df, aes_string(x = params$xColumn, y = params$yColumn)) +
          geom_point(size = 3, alpha = 0.7, color = "#667eea") +
          labs(title = params$title, x = params$xColumn, y = params$yColumn)
      }
      
    } else if (chart_type == "line") {
      # 3. LINE CHART - Time series and trends
      if (!is.null(params$groupColumn) && params$groupColumn != "") {
        p <- ggplot(df, aes_string(x = params$xColumn, y = params$yColumn, color = params$groupColumn, group = params$groupColumn)) +
          geom_line(size = 1.2) +
          geom_point(size = 2) +
          scale_color_viridis(discrete = TRUE) +
          labs(title = params$title, x = params$xColumn, y = params$yColumn)
      } else {
        p <- ggplot(df, aes_string(x = params$xColumn, y = params$yColumn)) +
          geom_line(size = 1.2, color = "#667eea") +
          geom_point(size = 2, color = "#f093fb") +
          labs(title = params$title, x = params$xColumn, y = params$yColumn)
      }
      
    } else if (chart_type == "bar") {
      # 4. BAR CHART - Categorical comparisons
      if (!is.null(params$groupColumn) && params$groupColumn != "") {
        p <- ggplot(df, aes_string(x = params$xColumn, y = params$yColumn, fill = params$groupColumn)) +
          geom_bar(stat = "identity", position = "dodge") +
          scale_fill_viridis(discrete = TRUE) +
          labs(title = params$title, x = params$xColumn, y = params$yColumn) +
          theme(axis.text.x = element_text(angle = 45, hjust = 1))
      } else {
        p <- ggplot(df, aes_string(x = params$xColumn, y = params$yColumn)) +
          geom_bar(stat = "identity", fill = "#667eea", alpha = 0.8) +
          labs(title = params$title, x = params$xColumn, y = params$yColumn) +
          theme(axis.text.x = element_text(angle = 45, hjust = 1))
      }
      
    } else if (chart_type == "histogram") {
      # 5. HISTOGRAM - Frequency distribution
      p <- ggplot(df, aes_string(x = params$xColumn)) +
        geom_histogram(bins = 30, fill = "#667eea", color = "#f093fb", alpha = 0.7) +
        labs(title = params$title, x = params$xColumn, y = "Frequency")
      
    } else if (chart_type == "violin") {
      # 6. VIOLIN PLOT - Distribution density
      p <- ggplot(df, aes_string(x = params$xColumn, y = params$yColumn)) +
        geom_violin(fill = "#667eea", alpha = 0.7) +
        geom_boxplot(width = 0.1, fill = "#f093fb", alpha = 0.5) +
        labs(title = params$title, x = params$xColumn, y = params$yColumn) +
        theme(axis.text.x = element_text(angle = 45, hjust = 1))
      
    } else if (chart_type == "density") {
      # 7. DENSITY PLOT - Continuous probability distribution
      if (!is.null(params$groupColumn) && params$groupColumn != "") {
        p <- ggplot(df, aes_string(x = params$xColumn, fill = params$groupColumn)) +
          geom_density(alpha = 0.6) +
          scale_fill_viridis(discrete = TRUE) +
          labs(title = params$title, x = params$xColumn, y = "Density")
      } else {
        p <- ggplot(df, aes_string(x = params$xColumn)) +
          geom_density(fill = "#667eea", alpha = 0.7) +
          labs(title = params$title, x = params$xColumn, y = "Density")
      }
      
    } else if (chart_type == "heatmap") {
      # 8. HEATMAP - Correlation matrix
      # Select only numeric columns
      numeric_cols <- df %>% select_if(is.numeric)
      cor_matrix <- cor(numeric_cols, use = "complete.obs")
      melted_cor <- melt(cor_matrix)
      
      p <- ggplot(melted_cor, aes(Var1, Var2, fill = value)) +
        geom_tile() +
        scale_fill_gradient2(low = "#667eea", mid = "white", high = "#f093fb", 
                            midpoint = 0, limit = c(-1, 1)) +
        labs(title = params$title, x = "", y = "", fill = "Correlation") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1))
      
    } else if (chart_type == "ridgeline") {
      # 9. RIDGELINE PLOT - Multiple distribution comparison
      p <- ggplot(df, aes_string(x = params$yColumn, y = params$xColumn, fill = params$xColumn)) +
        geom_density_ridges(alpha = 0.7) +
        scale_fill_viridis(discrete = TRUE) +
        labs(title = params$title, x = params$yColumn, y = params$xColumn) +
        theme(legend.position = "none")
      
    } else if (chart_type == "pie") {
      # 10. PIE CHART - Proportions and percentages
      # Aggregate data
      pie_data <- df %>%
        group_by(!!sym(params$xColumn)) %>%
        summarise(count = n()) %>%
        mutate(percentage = count / sum(count) * 100)
      
      p <- ggplot(pie_data, aes(x = "", y = count, fill = !!sym(params$xColumn))) +
        geom_bar(stat = "identity", width = 1) +
        coord_polar("y", start = 0) +
        scale_fill_viridis(discrete = TRUE) +
        labs(title = params$title, fill = params$xColumn) +
        theme_void() +
        theme(plot.title = element_text(size = 16, face = "bold", hjust = 0.5))
      
    } else {
      stop(paste("Unknown chart type:", chart_type))
    }
    
    # Save the plot
    ggsave(params$outputPath, plot = p, width = 10, height = 6, dpi = 300, bg = "white")
    cat("Chart created successfully:", params$outputPath, "\n")
    
  }, error = function(e) {
    cat("Error creating chart:", e$message, "\n")
    stop(e)
  })
}

# Create the chart
create_chart(params$chartType, df, params)
